File: 16. 3Sum Closest/16. 3Sum Closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int closestSum = nums[0] + nums[1] + nums[2];
        
        for (int i = 0; i < nums.size() - 2; i++) {
            int left = i + 1;
            int right = nums.size() - 1;
            
            while (left < right) {
                int currentSum = nums[i] + nums[left] + nums[right];
                
                if (abs(target - currentSum) < abs(target - closestSum)) {
                    closestSum = currentSum;
                }
                
                if (currentSum > target) {
                    right--;
                } else if (currentSum < target) {
                    left++;
                } else {
                    return target;  // Found exact sum
                }
            }
        }
        
        return closestSum;
    }
};

File: 16. 3Sum Closest/16. 3Sum Closest.rs
impl Solution {
    pub fn three_sum_closest(nums: Vec<i32>, target: i32) -> i32 {
        let mut nums = nums;
        nums.sort_unstable();
        let mut closest_sum = nums[0] + nums[1] + nums[2];
        
        for i in 0..nums.len() - 2 {
            let mut left = i + 1;
            let mut right = nums.len() - 1;
            
            while left < right {
                let current_sum = nums[i] + nums[left] + nums[right];
                
                if (target - current_sum).abs() < (target - closest_sum).abs() {
                    closest_sum = current_sum;
                }
                
                if current_sum > target {
                    right -= 1;
                } else if current_sum < target {
                    left += 1;
                } else {
                    return target;  // Found exact sum
                }
            }
        }
        
        closest_sum
    }
}